### DISCUSSION TITLE
Hvordan ønsker vi at attributter skal sendes fra EPJ til HelseID?

### DISCUSSION THREAD

**Discussion by runegri** [NorskHelsenett/Tillitsrammeverk#29] [General]
I en pull request (https://github.com/NorskHelsenett/Tillitsrammeverk/pull/25) fjernet jeg beskrivelsen av funsjonalitet som lar klientsystemet sende attributter til HelseID via token-endepunktet. Årsaken er at jeg mener vi går utover protokollspesifikasjonen ved å gjøre dette.
Samtidig så ser vi at å sende informasjon via token-endepunktet gir en mye bedre brukeropplevelse samtidig som det er mye enklere for systemleverandørene å implementere. Uten denne muligheten må i praksis brukeren gjøre en ny innlogging (men via SSO) mot HelseID for hver gang de endrer pasientkontekst. Jeg mener at dette blir unødvendig tungvindt. Å sende informasjon via authorize-endepunktet er nyttig dersom brukeren må gi et samtykke til operasjonen som gjennomføres, men i HelseID antar vi et implisitt samtykke og hopper derfor over denne funksjonen. 

Hva tenker dere andre om dette? Er særlig interessert i hvordan leverandørene ser på dette.

**Comment by dag-at-udelt** [comment]
Årsaken til at attributter er begrenset til å sendes frontend i OAuth2 er at det normalt forventes en "consent" fra sluttbruker når attributtene endres.

I vår sektor er consent implisitt, og da er det ingen god grunn til å ikke tillate endring av attributter uten å informere sluttbruker.

Prisen for å ikke godta oppdatering av tilstand backend (altså mot token endepunktet i HelseID) er:
- Dårligere brukeropplevelse, spesielt i webbaserte systemer
- Høyere teknisk kompleksitet for leverandører
- Høyere teknisk kompleksitet i HelseID (ettersom vi tillater oppdatering fra backend i andre scenarioer)

**Reply by runegri** [reply]
Helt enig, men savner kommentarer fra andre aktører som ikke er involvert i HelseID. Er dette er kurant tema å ta opp i et fellesmøte @steinarnoem ?

**Comment by troelde** [comment]
I EPJ-systemer med "embedded browser"-komponent, så kan det nok være mulig å skjule flimring som følge av SSO, men det må testes ut.  Vi må spørre oss om det er noen sikkerhetsmessige grunner til ikke å tilby direkte på API /token-endepunktet når det ikke kreves samtykke? Blir det ikke det samme som dersom klienten allerede har fått `offline_access` scope-tilgang med refresh token? For å overholde standard, så kan kanskje dette også tilbys på standard måte via /authorize-endepunktet?

**Comment by frodegil** [comment]
Hei

Helse-ID bør være en tjeneste for å etablere identitet - altså HVEM du er og eventuelt HVOR du er.

Andre attributter må ligge i datakontraktene mellom tjenesteleverandøren og konsumenten og ikke som logikk/claims i Helse-Id.

For eksempel: For pasientens samtykke burde det utvikles en egen tjeneste og portal der pasienten selv kan se og fjerne inngåtte samtykker. Tjenesteleverandøren kan bruke en slik tjeneste til å "dobbeltsjekke" at samtykket i API-kontrakten er korrekt.

Eksempel på mulig flyt i EPJ dersom samtykke ikke allerede registrert mellom behandler og pasient;

1) Pasienten gir legen muntlig samtykke
2) Legen registrerer samtykke via EPJ og sender denne elektronisk til samtykke-tjenesten. Tjenesten svarer med en beskrivelse av omfanget til samtykket (som legen leser opp for pasienten).
3) Endelig bekreftelse på samtykket registreres i tjenesten.
4) Pasienten mottar SMS og kan åpne samtykke portalen (Helse-Norge) for å sjekke at samtykket er korrekt registrert og har mulighet til å lese beskrivelsen av omfanget (som f.eks. at låste resepter vil kunne lastes ned fra KJ)
4) Legen henter pasientdata fra Kjernejournal via EPJ

Samtykke-tjenesten bør fungere litt som TTP (tiltrodd tredje-part) ala public-key insfrastrukturer.

Vi mener derfor at det finnes bedre måter å løse API-spesifikke utfordringer på enn å prøve å bake dette inn i Helse-ID logikk/claims.

En challenge-flyt fra en web-basert EPJ til Helse-Norge hver gang man bytter pasient vil gi en svært dårlig brukeropplevelse. 
Vær så snill og ikke gå den veien.

Mvh Frode Gilberg
Utviklingssjef PRIDOK
